### PR TITLE
Vulkan nv12

### PR DESCRIPTION
**Pull request by wenbinc-Bin**

### CODE REVIEW

**Comment by wenbinc-Bin** [comment]
@xhaihao I try to fix nv12 and nv21 problem like this. Any suggestion? If no I will send to community.
Thanks
Wenbin

**Comment by xhaihao** [comment]
How about change the commit log in the 2nd commit to: "The UV plane of NV21 should be mapped to GR88" ? We can't map GR88 to NV21.

**Comment by wenbinc-Bin** [comment]
Thanks for review. I will update it.

File: libavutil/hwcontext_vulkan.c [hunk 1]
```diff
@@ -3290,8 +3290,13 @@ static void vulkan_unmap_to_drm(AVHWFramesContext *hwfc, HWMapDescriptor *hwmap)
...
 }
 
-static inline uint32_t vulkan_fmt_to_drm(VkFormat vkfmt)
+static inline uint32_t vulkan_fmt_to_drm(VkFormat vkfmt,
+                                         enum AVPixelFormat sw_format)
 {
+    if (vkfmt == VK_FORMAT_R8G8_UNORM &&
+       (sw_format == AV_PIX_FMT_NV21 || sw_format == AV_PIX_FMT_NV42))
+            return DRM_FORMAT_GR88;
+
```
Comment (xhaihao) [inline libavutil/hwcontext_vulkan.c:3303]: How about using code below ? If so, we needn't swap the positions in the list

```
if (vkfmt == VK_FORMAT_R8G8_UNORM &&
       (sw_format == AV_PIX_FMT_NV12 || sw_format == AV_PIX_FMT_NV24))
            return DRM_FORMAT_RG88;
```

Comment (xhaihao) [reply libavutil/hwcontext_vulkan.c:3303]: Is a similar check needed for `VK_FORMAT_R16G16_UNORM` ?

File: libavutil/hwcontext_vaapi.c [hunk 2]
```diff
@@ -992,7 +992,7 @@ static const struct {
 } vaapi_drm_format_map[] = {
 #ifdef DRM_FORMAT_R8
     DRM_MAP(NV12, 2, DRM_FORMAT_R8,  DRM_FORMAT_RG88),
-    DRM_MAP(NV12, 2, DRM_FORMAT_R8,  DRM_FORMAT_GR88),
+    DRM_MAP(NV21, 2, DRM_FORMAT_R8,  DRM_FORMAT_GR88),
```
Comment (xhaihao) [inline libavutil/hwcontext_vaapi.c]: Mapping UV plane of NV12 to DRM_FORMAT_GR88 is WA, and I think we needn't add support for NV21 and  should remove `DRM_MAP(NV12, 2, DRM_FORMAT_R8,  DRM_FORMAT_GR88)` after applying the first commit.

BTW so far we don't have a driver to support NV21

File: libavutil/hwcontext_vulkan.c (hunk 1 above)
Comment (wenbinc-Bin) [reply libavutil/hwcontext_vulkan.c:3303]: VK_FORMAT_R16G16_UNORM will be mapped to DRM_FORMAT_GR1616 which is wrong so I swap it's position and I swap R8G8 as well.
I see there are only P010 and P016 using VK_FORMAT_R16G16_UNORM, and DRM_FORMAT_GR1616 is not used, so I think we don't need to check R16G16.

File: libavutil/hwcontext_vaapi.c (hunk 2 above)
Comment (wenbinc-Bin) [reply libavutil/hwcontext_vaapi.c]: Ok, I will delete this line.

File: libavutil/hwcontext_vulkan.c (hunk 1 above)
Comment (xhaihao) [reply libavutil/hwcontext_vulkan.c:3303]: If so, how about to remove `{ DRM_FORMAT_GR1616,   VK_FORMAT_R16G16_UNORM   },` from this list ?

**Comment by xhaihao** [comment]
typo in commit log:
s/prblem/problem

**Comment by xhaihao** [comment]
You fixed both r8g8 and r16g16, could you update the commit log of the first commit ?

**Comment by wenbinc-Bin** [comment]
The commit log is updated.

**Review by xhaihao — APPROVED** [review]

**Comment by wenbinc-Bin** [comment]
submit to: https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=6329

**Pull request closed without merging**